src/systems/buoyancy_engine/BuoyancyEngine.hh

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (C) 2019 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#ifndef IGNITION_GAZEBO_SYSTEMS_BALLAST_HH_
+#define IGNITION_GAZEBO_SYSTEMS_BALLAST_HH_
+
+#include <ignition/gazebo/System.hh>
+
+#include <memory>
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace systems
+{
+  class BuoyancyEnginePrivateData;
+
+  /// \brief This class provides a simple mechanical bladder which is used to
+  /// control the buoyancy of an underwater glider. It uses archimedes principal
+  /// to apply an upward force based on the volume of the bladder. It listens to
+  /// the topic `buoyancy_engine` or `/model/{namespace}/buoyancy_engine`
+  /// topic for the volume of the bladder in *cubicmeters*.
+  ///
+  /// ## Parameters
+  /// <link_name> - The link which the plugin is attached to [required, string]
+  /// <namespace> - The namespace for the topic. If empty the plugin will listen
+  ///   on `buoyancy_engine` otherwise it listens on
+  ///   `/model/{namespace}/buoyancy_engine` [optional, string]
+  /// <min_volume> - Minimum volume of the engine [optional, float,
+  ///   default=0.00003m^3]
+  /// <neutral_volume> - At this volume the engine has neutral buoyancy. Used to
+  ///   estimate the weight of the engine [optional, float, default=0.0003m^3]
+  /// <default_volume> - The volume which the engine starts at [optional, float,
+  ///   default=0.0003m^3]
+  /// <max_volume> - Maximum volume of the engine [optional, float,
+  ///   default=0.00099m^3]
+  /// <max_inflation_rate> - Maximum inflation rate for bladder [optional,
+  ///   float, default=0.000003m^3/s]
+  ///
```

Comment:
Missing `<fluid_density>`. And I'm curious about whether you've thought about how this interacts with the layered buoyancy being added on #818 .

Reply:
Resolved in 4b96340

With regards to #818. You're right about it being incompatible particularly if the ballast volume is comparable to the size of the vessel itself. I personally feel that this is something that we can fix later. We also should consider how a user actually sets up the fluid density of worlds. Currently, we define this within the Buoyancy plugin and this plugin separately. Perhaps in the future we can have some way of specifying these properties at the world level.